storm-core/src/jvm/org/apache/storm/daemon/supervisor/ReadClusterState.java

```diff
@@ -0,0 +1,318 @@
+import org.apache.storm.cluster.IStormClusterState;
+import org.apache.storm.cluster.VersionedData;
+import org.apache.storm.daemon.supervisor.Slot.MachineState;
+import org.apache.storm.daemon.supervisor.Slot.TopoProfileAction;
+import org.apache.storm.event.EventManager;
+import org.apache.storm.generated.Assignment;
+import org.apache.storm.generated.ExecutorInfo;
+import org.apache.storm.generated.LocalAssignment;
+import org.apache.storm.generated.NodeInfo;
+import org.apache.storm.generated.ProfileRequest;
+import org.apache.storm.generated.WorkerResources;
+import org.apache.storm.localizer.ILocalizer;
+import org.apache.storm.messaging.IContext;
+import org.apache.storm.scheduler.ISupervisor;
+import org.apache.storm.utils.LocalState;
+import org.apache.storm.utils.Time;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ReadClusterState implements Runnable, AutoCloseable {
+    private static final Logger LOG = LoggerFactory.getLogger(ReadClusterState.class);
+    
+    private final Map<String, Object> superConf;
+    private final IStormClusterState stormClusterState;
+    private final EventManager syncSupEventManager;
+    private final AtomicReference<Map<String, VersionedData<Assignment>>> assignmentVersions;
+    private final Map<Integer, Slot> slots = new HashMap<>();
+    private final AtomicInteger readRetry = new AtomicInteger(0);
+    private final String assignmentId;
+    private final ISupervisor iSuper;
+    private final ILocalizer localizer;
+    private final ContainerLauncher launcher;
+    private final String host;
+    private final LocalState localState;
+    private final IStormClusterState clusterState;
+    private final AtomicReference<Map<Long, LocalAssignment>> cachedAssignments;
+    
+    public ReadClusterState(Supervisor supervisor) throws Exception {
+        this(supervisor.getConf(), supervisor.getStormClusterState(), supervisor.getEventManger(),
+                supervisor.getAssignmentId(), supervisor.getiSupervisor(),
+                supervisor.getAsyncLocalizer(), supervisor.getHostName(),
+                supervisor.getLocalState(), supervisor.getStormClusterState(),
+                supervisor.getCurrAssignment(), supervisor.getSharedContext());
+    }
+    
+    public ReadClusterState(Map<String, Object> superConf, IStormClusterState stormClusterState,
+            EventManager syncSupEventManager, String assignmentId, ISupervisor iSuper,
+            ILocalizer localizer, String host, LocalState localState,
+            IStormClusterState clusterState, AtomicReference<Map<Long, LocalAssignment>> cachedAssignments,
+            IContext sharedContext) throws Exception{
+        this.superConf = superConf;
+        this.stormClusterState = stormClusterState;
+        this.syncSupEventManager = syncSupEventManager;
+        this.assignmentVersions = new AtomicReference<Map<String, VersionedData<Assignment>>>(new HashMap<String, VersionedData<Assignment>>());
```

Comment:
I tried that, and for some reason java doesn't like it, because I am initializing it with a HashMap, that is a Map, but not exactly a Map. I wasn't totally sure why java didn't like it.